README.md
# DEPRECATED

This module is no longer maintained by StrongLoop/IBM.

We are encouraging the community to take the ownership of this module.
Please open an issue in https://github.com/strongloop/loopback if you
are willing to become the new maintainer.

See also https://groups.google.com/forum/#!topic/loopbackjs/w038RvqHeYI

---

#loopback-testing

**Utilities for testing LoopBack apps**

## overview

The following helpers are designed to generate [mocha] tests against
[LoopBack](http://strongloop.com/loopback) apps.

## install

1. `npm install loopback-testing --save-dev`
2. Assuming you started with a clean template/project generated by `slc loopback`
  1. If you have mocha installed as a global npm module that's great! Simply update `<your_project>/package.json` with:

    ```
    {
      ...
      "scripts": {
        ...
        "test": "mocha"
      }
    }
    ```
  2. Otherwise, you can utilize the mocha library within the `loopback-testing` testing module:

    ```
    {
      ...
      "scripts": {
        ...
        "test": "./node_modules/loopback-testing/node_modules/.bin/mocha"
      }
    }
    ```
3. Run `npm test` to execute any tests under the `test` directory.

## basic usage

Below is a simple LoopBack app.

```js
var loopback = require('loopback');
var app = loopback();
var Product = app.model('product');
Product.attachTo(loopback.memory());
```

Use the `loopback-testing` module to generate `mocha` tests.

```js
var lt = require('loopback-testing');
var assert = require('assert');
var app = require('../server/server.js'); //path to app.js or server.js

describe('/products', function() {
  lt.beforeEach.withApp(app);
  lt.describe.whenCalledRemotely('GET', '/products', function() {
    lt.it.shouldBeAllowed();
    it('should have statusCode 200', function() {
      assert.equal(this.res.statusCode, 200);
    });

    lt.beforeEach.givenModel('product');
    it('should respond with an array of products', function() {
      assert(Array.isArray(this.res.body));
    });
  });
});
```

## building test data

Use TestDataBuilder to build many Model instances in one async call. Specify
only properties relevant to your test, the builder will pre-fill remaining
required properties with sensible defaults.

```js
var TestDataBuilder = require('loopback-testing').TestDataBuilder;
var ref = TestDataBuilder.ref;

// The context object to hold the created models.
// You can use `this` in mocha test instead.
var context = {};

var ref = TestDataBuilder.ref;
new TestDataBuilder()
  .define('application', Application, {
    pushSettings: { stub: { } }
  })
  .define('device', Device, {
     // use the value of application's id
     // the value is resolved at build time
     appId: ref('application.id'),
     deviceType: 'android'
  })
  .define('notification', Notification)
  .buildTo(context, function(err) {
    // test models are available as
    //   context.application
    //   context.device
    //   context.notification
  });
```


CHANGES.md
2016-09-12, Version 1.4.0
=========================

 * Update dependencies to the latest version (Miroslav Bajtoš)


2016-08-10, Version 1.3.0
=========================

 * Update deprecation notice (Miroslav Bajtoš)

 * Update URLs in CONTRIBUTING.md (#73) (Ryan Graham)

 * update copyright notices and license (Ryan Graham)

 * Updates (Simon Ho)

 * Update README.md (Simon Ho)

 * Refer to licenses with a link (Sam Roberts)

 * Use strongloop conventions for licensing (Sam Roberts)

 * Move mocha to dev deps (Ritchie Martori)

 * Wait for automigrate to finish (Vlad Miller)

 * Replace call of login method on constructor (Aris Kemper)


2015-06-24, Version 1.2.0
=========================

 * Respect model property length for TDB strings (zero5100)

 * use object to support placeholder pattern for when called remotely (ningsuhen)

 * use findorCreate to create roles (ningsuhen)

 * Initial withUserModel feature (Ryan Schumacher)

 * Update helpers.js (Vladyslav Millier)


2015-02-23, Version 1.1.0
=========================

 * Add done callback (Ryan Schumacher)

 * Add clean datasource feature (Ryan Schumacher)


2015-01-13, Version 1.0.5
=========================

 * Throw err when beforeEach.withApp is not called (Farid Neshat)

 * Fix bad CLA URL in CONTRIBUTING.md (Ryan Graham)


2014-12-05, Version 1.0.4
=========================

 * README: add basic mocha setup instructions (pulkitsinghal)


2014-10-02, Version 1.0.3
=========================

 * Update version to 1.0.3 (Simon Ho)


2014-10-02, Version 1.0.2
=========================



2014-10-02, Version 1.0.1
=========================

 * package: move chai from devDependencies to dependencies (Simon Ho)

 * Add contribution guidelines (Ryan Graham)


2014-09-26, Version 1.0.0
=========================

 * Upgrade dependencies (Miroslav Bajtoš)

 * add support for testing with users with roles (Jaka Hudoklin)

 * fix imports (Jaka Hudoklin)


2014-09-25, Version 0.2.1
=========================

 * Fix handling of dynamic URLs in the test builder (Clark Wang)

 * Extending tests to cover other methods (ariskemper)

 * package: add repository URL (Miroslav Bajtoš)


2014-06-12, Version 0.2.0
=========================

 * helpers: add parameter for request body (Miroslav Bajtoš)

 * helpers: Improve ACL asserts (Miroslav Bajtoš)

 * helpers: fix usage of mocha contexts (Miroslav Bajtoš)

 * Drop peer dependency on loopback (Miroslav Bajtoš)

 * un-used variable (Karl Mikkelsen)

 * tighter status constraints (Karl Mikkelsen)

 * extend range of status allowed/denied (Karl Mikkelsen)


2014-06-06, Version 0.1.5
=========================

 * Support loopback 2.x (Miroslav Bajtoš)

 * Bump loopback dependency (Ritchie Martori)


2014-05-25, Version 0.1.3
=========================

 * Remove juggler from peerDependencies. (Miroslav Bajtoš)

 * Update to dual MIT/StrongLoop license (Raymond Feng)


2014-02-11, Version 0.1.2
=========================

 * Bump version and update deps (Raymond Feng)


2014-01-13, Version 0.1.1
=========================

 * Speed up login action (Miroslav Bajtoš)

 * Fix function typo (Ritchie Martori)


2013-12-20, Version 0.1.0
=========================

 * Relax loopback dependency (Ritchie Martori)

 * Factor out LoopBack dep to removing coupling on the exact loopback instance. (Ritchie Martori)

 * Remove describe.model (Ritchie Martori)

 * Remove beforeEach.withDefaultDataSource (Ritchie Martori)

 * Move helpers to lib (Ritchie Martori)


2013-12-13, Version 0.0.4
=========================

 * Bump version (Ritchie Martori)

 * Move loopback to peerDependencies to avoid using the wrong loopback object (Ritchie Martori)

 * Describe TestDataBuilder in README. (Miroslav Bajtos)


2013-12-11, Version 0.0.3
=========================

 * First release!


test/test.js
// Copyright IBM Corp. 2013,2015. All Rights Reserved.
// Node module: loopback-testing
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

var loopback = require('loopback');
var helpers = require('../');
var assert = require('assert');

describe('helpers', function () {
  var testApp = loopback();
  var db = testApp.dataSource('db', {connector: loopback.Memory});
  var testModel = testApp.model('xxx-test-model', {dataSource: 'db'});

  testApp.use(loopback.rest());
  helpers.beforeEach.withApp(testApp);

  describe('helpers.it', function() {
    ['shouldBeAllowed',
     'shouldBeDenied',
     'shouldBeRejected',
     'shouldNotBeFound',
     'shouldBeAllowedWhenCalledAnonymously',
     'shouldBeDeniedWhenCalledAnonymously',
     'shouldBeAllowedWhenCalledUnauthenticated',
     'shouldBeDeniedWhenCalledUnauthenticated',
     'shouldBeNotFoundWhenCalledUnauthenticated',
     'shouldBeAllowedWhenCalledByUser',
     'shouldBeValidCreateResponse',
     'shouldBeValidGetAllResponse',
     'shouldBeValidGetByIdResponse',
     'shouldBeValidUpdateResponse',
     'shouldBeValidDeleteResponse',
     'shouldBeDeniedWhenCalledByUser']
    .forEach(function(func) {
      it('should have a method named ' + func, function () {
        assert.equal(typeof helpers.it[func], 'function');
      });
    });
  });

  describe('helpers.describe', function() {
    ['staticMethod',
     'instanceMethod',
     'whenLoggedInAsUser',
     'whenCalledByUser',
     'whenCalledAnonymously',
     'whenCalledUnauthenticated']
    .forEach(function(func) {
      it('should have a method named ' + func, function () {
        assert.equal(typeof helpers.describe[func], 'function');
      });
    });
  });

  describe('helpers.beforeEach', function() {
    ['withArgs',
     'givenModel',
     'givenUser',
     'givenLoggedInUser',
     'givenAnUnauthenticatedToken',
     'givenAnAnonymousToken']
    .forEach(function(func) {
      it('should have a helper method named ' + func, function () {
        assert.equal(typeof helpers.beforeEach[func], 'function');
      });
    });
  });

  describe('helpers.beforeEach.givenModel', function() {
    helpers.beforeEach.givenModel('xxx-test-model');
    it('should have an xxx-test-model property', function () {
      assert(this['xxx-test-model']);
      assert(this['xxx-test-model'].id);
    });
  });

  describe('whenCalledRemotely', function() {
    helpers.describe.staticMethod('create', function() {
      helpers.beforeEach.withArgs({foo: 'bar'});
      helpers.describe.whenCalledRemotely('POST', '/xxx-test-models', function() {
        it('should call the method over rest', function () {
          assert.equal(this.res.statusCode, 200);
        });
      });
    });
    helpers.describe.staticMethod('findById', function() {
      helpers.beforeEach.givenModel('xxx-test-model', {foo: 'bar'});
      helpers.describe.whenCalledRemotely('GET', function () {
        return '/xxx-test-models/' + this['xxx-test-model'].id;
      }, function() {
        it('should retrieve the expected model in the first test', function () {
          assert.equal(this.res.body.id, this['xxx-test-model'].id);
        });
        it('should retrieve the expected model in subsequent tests', function () {
          assert.equal(this.res.body.id, this['xxx-test-model'].id);
        });
      });
    });
  });

  describe('cleanDatasource', function() {
    helpers.describe.staticMethod('create', function() {
      helpers.beforeEach.withArgs({foo: 'bar'});
      helpers.describe.whenCalledRemotely('POST', '/xxx-test-models', function() {
        it('should call the method over rest', function () {
          assert.equal(this.res.statusCode, 200);
        });
      });
    });

    helpers.describe.staticMethod('findById', function() {
      helpers.beforeEach.givenModel('xxx-test-model', {foo: 'bar'});
      helpers.beforeEach.cleanDatasource();
      helpers.describe.whenCalledRemotely('GET', function () {
        return '/xxx-test-models/' + this['xxx-test-model'].id;
      }, function() {
        it('should not find the given model', function () {
          assert.equal(this.res.statusCode, 404);
        });
      });
    });
  });
});


test/test-data-builder.test.js
// Copyright IBM Corp. 2013,2015. All Rights Reserved.
// Node module: loopback-testing
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

var loopback = require('loopback');
var TestDataBuilder = require('../lib/test-data-builder');
var expect = require('chai').expect;

describe('TestDataBuilder', function() {
  var db;
  var TestModel;

  beforeEach(function() {
    db = loopback.createDataSource({ connector: loopback.Memory });
  });

  it('builds a model', function(done) {
    givenTestModel({ value: String });

    new TestDataBuilder()
      .define('model', TestModel, { value: 'a-string-value' })
      .buildTo(this, function(err) {
        if (err) return done(err);
        expect(this.model).to.have.property('value', 'a-string-value');
        done();
      }.bind(this));
  });

  // Parameterized test
  function itAutoFillsRequiredPropertiesWithUniqueValuesFor(type) {
    it(
      'auto-fills required ' + type + ' properties with unique values',
      function(done) {
        givenTestModel({
          required1: { type: type, required: true },
          required2: { type: type, required: true }
        });

        new TestDataBuilder()
          .define('model', TestModel, {})
          .buildTo(this, function(err) {
            if (err) return done(err);
            expect(this.model.required1).to.not.equal(this.model.required2);
            expect(this.model.optional).to.satisfy(notSet);
            done();
          }.bind(this));
      }
    );
  }

  itAutoFillsRequiredPropertiesWithUniqueValuesFor(String);
  itAutoFillsRequiredPropertiesWithUniqueValuesFor(Number);
  itAutoFillsRequiredPropertiesWithUniqueValuesFor(Date);

  it('limits the length of the generated string value to the property length', function(done) {
    var testMaxStringLength = 10;
    givenTestModel({
      required1: { type: String, required: true },
      required2: { type: String, required: true, length: testMaxStringLength }
    });

    new TestDataBuilder()
      .define('model', TestModel, {})
      .buildTo(this, function(err) {
        if (err) return done(err);
        expect(this.model.required1).to.not.equal(this.model.required2);
        expect(this.model.required2).to.have.length.of.at.most(testMaxStringLength);
        done();
      }.bind(this));
  });

  it('auto-fills required Boolean properties with false', function(done) {
    givenTestModel({
      required: { type: Boolean, required: true }
    });

    new TestDataBuilder()
      .define('model', TestModel, {})
      .buildTo(this, function(err) {
        if (err) return done(err);
        expect(this.model.required).to.equal(false);
        done();
      }.bind(this));
  });

  it('does not fill optional properties', function(done) {
    givenTestModel({
      optional: { type: String, required: false }
    });

    new TestDataBuilder()
      .define('model', TestModel, {})
      .buildTo(this, function(err) {
        if (err) return done(err);
        expect(this.model.optional).to.satisfy(notSet);
        done();
      }.bind(this));
  });

  it('resolves references', function(done) {
    var Parent = givenModel('Parent', { name: { type: String, required: true } });
    var Child = givenModel('Child', { parentName: String });

    new TestDataBuilder()
      .define('parent', Parent)
      .define('child', Child, {
        parentName: TestDataBuilder.ref('parent.name')
      })
      .buildTo(this, function(err) {
        if(err) return done(err);
        expect(this.child.parentName).to.equal(this.parent.name);
        done();
      }.bind(this));
  });

  function givenTestModel(properties) {
    TestModel = givenModel('TestModel', properties);
  }

  function givenModel(name, properties) {
    var ModelCtor = loopback.createModel(name, properties);
    ModelCtor.attachTo(db);
    return ModelCtor;
  }

  function notSet(value) {
    // workaround for `expect().to.exist` that triggers a JSHint error
    // (a no-op statement discarding the property value)
    return value === undefined || value === null;
  }
});


lib/request-builder.js
var extend = require('util')._extend;
var async = require('async');

module.exports = exports = TestFLow;

/**
 * Make HTTP requests against loopback apps.
 *
 * Usage:
 * ```js

 * ```
 * @constructor
 */
function TestFLow() {
  this._definitions = [];
}

/**
 * Define a new model instance.
 * @param {string} name Name of the instance.
 *   `buildTo()` will save the instance created as context[name].
 * @param {constructor} Model Model class/constructor.
 * @param {Object.<string, Object>=} properties
 *   Properties to set in the object.
 *   Intelligent default values are supplied by the builder
 *   for required properties not listed.
 * @return TestFLow (fluent interface)
 */
TestFLow.prototype.define = function(name, Model, properties) {
  this._definitions.push({
    name: name,
    model: Model,
    properties: properties
  });
  return this;
};

/**
 * Reference the value of a property from a model instance defined before.
 * @param {string} path Generally in the form '{name}.{property}', where {name}
 * is the name passed to `define()` and {property} is the name of
 * the property to use.
 */
TestFLow.ref = function(path) {
  return new Reference(path);
};

/**
 * Asynchronously build all models defined via `define()` and save them in
 * the supplied context object.
 * @param {Object.<string, Object>} context The context to object to populate.
 * @param {function(Error)} callback Callback.
 */
TestFLow.prototype.buildTo = function(context, callback) {
  this._context = context;
  async.eachSeries(
    this._definitions,
    this._buildObject.bind(this),
    callback);
};

TestFLow.prototype._buildObject = function(definition, callback) {
  var defaultValues = this._gatherDefaultPropertyValues(definition.model);
  var values = extend(defaultValues, definition.properties || {});
  var resolvedValues = this._resolveValues(values);

  definition.model.create(resolvedValues, function(err, result) {
    if (err) {
      console.error(
        'Cannot build object %j - %s\nDetails: %j',
        definition,
        err.message,
        err.details);
    } else {
      this._context[definition.name] = result;
    }

    callback(err);
  }.bind(this));
};

TestFLow.prototype._resolveValues = function(values) {
  var result = {};
  for (var key in values) {
    var val = values[key];
    if (val instanceof Reference) {
      val = values[key].resolveFromContext(this._context);
    }
    result[key] = val;
  }
  return result;
};

var valueCounter = 0;
TestFLow.prototype._gatherDefaultPropertyValues = function(Model) {
  var result = {};
  Model.forEachProperty(function createDefaultPropertyValue(name) {
    var prop = Model.definition.properties[name];
    if (!prop.required) return;

    switch (prop.type) {
      case String:
        result[name] = 'a test ' + name + ' #' + (++valueCounter);
        break;
      case Number:
        result[name] = 1230000 + (++valueCounter);
        break;
      case Date:
        result[name] = new Date(
          2222, 12, 12, // yyyy, mm, dd
          12, 12, 12,   // hh, MM, ss
          ++valueCounter // milliseconds
        );
        break;
      case Boolean:
        // There isn't much choice here, is it?
        // Let's use "false" to encourage users to be explicit when they
        // require "true" to turn some flag/behaviour on
        result[name] = false;
        break;
      // TODO: support nested structures - array, object
    }
  });
  return result;
};

/**
 * Placeholder for values that will be resolved during build.
 * @param path
 * @constructor
 * @private
 */
function Reference(path) {
  this._path = path;
}

Reference.prototype.resolveFromContext = function(context) {
  var elements = this._path.split('.');

  var result = elements.reduce(
    function(obj, prop) {
      return obj[prop];
    },
    context
  );

  return result;
};


lib/helpers.js
// Copyright IBM Corp. 2013,2015. All Rights Reserved.
// Node module: loopback-testing
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

var _describe = {};
var _it = {};
var _beforeEach = {};
var helpers = exports = module.exports = {
  describe: _describe,
  it: _it,
  beforeEach: _beforeEach
};
var assert = require('assert');
var request = require('supertest');
var expect = require('chai').expect;

_beforeEach.withApp = function(app) {
  if (app.models.User) {
    // Speed up the password hashing algorithm
    app.models.User.settings.saltWorkFactor = 4;
  }

  beforeEach(function() {
    this.app = app;
    var _request = this.request = request(app);
    this.post = _request.post;
    this.get = _request.get;
    this.put = _request.put;
    this.del = _request.del;
  });
}

_beforeEach.cleanDatasource = function(dsName) {
  beforeEach(function(done) {
    if(!dsName) dsName = 'db';

    if (typeof this.app === 'function'
        && typeof this.app.datasources === 'object'
        && typeof this.app.datasources[dsName] === 'object') {
      this.app.datasources[dsName].automigrate(done);
      this.app.datasources[dsName].connector.ids = {};
    } else {
      done();
    }
  });
}

function mixin(obj, into) {
  Object.keys(obj).forEach(function(key) {
    if(typeof obj[key] === 'function') {
      into[key] = obj[key];
    }
  });
}

_describe.staticMethod = function(methodName, cb) {
  describe('.' + methodName, function() {
    beforeEach(function() {
      this.method = methodName;
      this.isStaticMethod = true;
    });
    cb();
  });
}

_describe.instanceMethod = function(methodName, cb) {
  describe('.prototype.' + methodName, function() {
    beforeEach(function() {
      this.method = methodName;
      this.isInstanceMethod = true;
    });
    cb();
  });
}

_beforeEach.withArgs = function() {
  var args = Array.prototype.slice.call(arguments, 0);
  beforeEach(function() {
    this.args = args;
  });
}

_beforeEach.givenModel = function(modelName, attrs, optionalHandler) {
  var modelKey = modelName;

  if(typeof attrs === 'function') {
    optionalHandler = attrs;
    attrs = undefined;
  }

  if(typeof optionalHandler === 'string') {
    modelKey = optionalHandler;
  }

  attrs = attrs || {};

  beforeEach(function(done) {
    if(modelName === '__USERMODEL__') {
      modelName = this.userModel ? this.userModel : 'user';
    }

    if(modelName === '__ACCESSTOKENMODEL__') {
      modelName = this.accessTokenModel ? this.accessTokenModel : 'accessToken';
    }

    if(modelName === '__ROLEMAPPINGMODEL__') {
      modelName = this.roleMappingModel ? this.roleMappingModel : 'roleMapping';
    }

    if(modelName === '__ROLEMODEL__') {
      modelName = this.roleModel ? this.roleModel : 'Role';
    }

    var test = this;
    var app = this.app;
    var model = app.models[modelName];
    assert(model, 'cannot get model of name ' + modelName + ' from app.models');
    assert(model.dataSource, 'cannot test model '+ modelName
      + ' without attached dataSource');
    assert(
      typeof model.create === 'function',
      modelName + ' does not have a create method'
    );

    model.create(attrs, function(err, result) {
      if(err) {
        console.error(err.message);
        console.error('Tried creating ' + modelName, attrs);
        if(err.details) console.error(err.details);
        done(err);
      } else {
        test[modelKey] = result;
        done();
      }
    });
  });

  if(typeof optionalHandler === 'function') {
    beforeEach(optionalHandler);
  }

  afterEach(function(done) {
    this[modelKey].destroy(done);
  });
}

_beforeEach.withUserModel = function(model) {
  beforeEach(function(done) {
    this.userModel = model;
    done();
  });
};

_beforeEach.withAccessTokenModel = function(model) {
  beforeEach(function(done) {
    this.accessTokenModel = model;
    done();
  });
};

_beforeEach.withRoleMappingModel = function(model) {
  beforeEach(function(done) {
    this.roleMappingModel = model;
    done();
  });
};

_beforeEach.withRoleModel = function(model) {
  beforeEach(function(done) {
    this.roleModel = model;
    done();
  });
};

_beforeEach.givenUser = function(attrs, optionalHandler) {
  _beforeEach.givenModel('__USERMODEL__', attrs, optionalHandler);
}

_beforeEach.givenUserWithRole = function (attrs, role, optionalHandler) {
  if (typeof role === 'string') {
    role = {
      name: role
    }
  }
  _beforeEach.givenUser(attrs, function (done) {
    var test = this;
    test.app.models[this.roleModel].findOrCreate({name: role}, function (err, result) {
      if(err) {
        console.error(err.message);
        if(err.details) console.error(err.details);
        return done(err);
      }

      test.userRole = result;
      test.app.models[this.roleMappingModel].create(
        {principalId: test.user.id,
         principalType: test.app.models[this.roleMappingModel].USER,
         roleId: result.id},
        function (err, result) {
          if(err) {
            console.error(err.message);
            if(err.details) console.error(err.details);
            return done(err);
          }

          test.userRoleMapping = result;
          done();
        }
      );
    });
  });

  if(typeof optionalHandler === 'function') {
    beforeEach(optionalHandler);
  }

  afterEach(function(done) {
    var test = this;
    this.userRole.destroy(function(err) {
      if(err) return done(err);
      test.userRole = undefined;

      test.userRoleMapping.destroy(function(err) {
        if(err) return done(err);
        test.userRoleMapping = undefined;
        done();
      });
    });
  });
}

_beforeEach.givenLoggedInUser = function(credentials, optionalHandler) {
  _beforeEach.givenUser(credentials, function(done) {
    var test = this;
    this.app.models[this.userModel].login(credentials, function(err, token) {
      if(err) {
        done(err);
      } else {
        test.loggedInAccessToken = token;
        done();
      }
    });
  });

  afterEach(function(done) {
    var test = this;
    this.loggedInAccessToken.destroy(function(err) {
      if(err) return done(err);
      test.loggedInAccessToken = undefined;
      done();
    });
  });
}

_beforeEach.givenLoggedInUserWithRole = function(credentials, role, optionalHandler){
  _beforeEach.givenUserWithRole(credentials, role, function(done) {
    var test = this;
    this.app.models[this.userModel].login(credentials, function(err, token) {
      if(err) {
        done(err);
      } else {
        test.loggedInAccessToken = token;
        done();
      }
    });
  });

  afterEach(function(done) {
    var test = this;
    this.loggedInAccessToken.destroy(function(err) {
      if(err) return done(err);
      test.loggedInAccessToken = undefined;
      done();
    });
  });
}

_beforeEach.givenAnUnauthenticatedToken = function(attrs, optionalHandler) {
  _beforeEach.givenModel('__ACCESSTOKENMODEL__', attrs, optionalHandler);
}

_beforeEach.givenAnAnonymousToken = function(attrs, optionalHandler) {
  _beforeEach.givenModel('__ACCESSTOKENMODEL__', {id: '$anonymous'}, optionalHandler);
}

_describe.whenCalledRemotely = function(verb, url, data, cb) {
  if (cb == undefined) {
    cb = data;
    data = null;
  }

  var urlStr = url;
  if(typeof url === 'function') {
    urlStr = '/<dynamic>';
  }
  else if(typeof url === 'object' && url.hasOwnProperty('placeHolder')) {
    urlStr = url.placeHolder;
  }

  describe(verb.toUpperCase() + ' ' + urlStr, function() {
    beforeEach(function(cb) {
      if(typeof url === 'function') {
        this.url = url.call(this);
      }
      else if(typeof url === 'object' && url.hasOwnProperty('callback')){
        this.url = url.callback.call(this);
      }
      this.remotely = true;
      this.verb = verb.toUpperCase();
      this.url = this.url || url;
      var methodForVerb = verb.toLowerCase();
      if(methodForVerb === 'delete') methodForVerb = 'del';

      if (this.request === undefined) {
          throw new Error('App is not specified. Please use lt.beforeEach.withApp to specify the app.');
      }

      this.http = this.request[methodForVerb](this.url);
      delete this.url;
      this.http.set('Accept', 'application/json');
      if(this.loggedInAccessToken) {
        this.http.set('authorization', this.loggedInAccessToken.id);
      }
      if (data) {
        var payload = data;
        if (typeof data === 'function')
          payload = data.call(this);
        this.http.send(payload);
      }
      this.req = this.http.req;
      var test = this;
      this.http.end(function(err) {
        test.req = test.http.req;
        test.res = test.http.res;
        delete test.url;
        cb();
      });
    });

    cb();
  });
}

_describe.whenLoggedInAsUser = function(credentials, cb) {
  describe('when logged in as user', function () {
    _beforeEach.givenLoggedInUser(credentials);
    cb();
  });
}

_describe.whenLoggedInAsUserWithRole = function(credentials, role, cb) {
  describe('when logged in as user', function () {
    _beforeEach.givenLoggedInUser(credentials, role);
    cb();
  });
}

_describe.whenCalledByUser = function(credentials, verb, url, data, cb) {
  describe('when called by logged in user', function () {
    _beforeEach.givenLoggedInUser(credentials);
    _describe.whenCalledRemotely(verb, url, data, cb);
  });
}

_describe.whenCalledByUserWithRole = function (credentials, role, verb, url, data, cb) {
   describe('when called by logged in user with role ' + role, function () {
    _beforeEach.givenLoggedInUserWithRole(credentials, role);
    _describe.whenCalledRemotely(verb, url, data, cb);
  });
}

_describe.whenCalledAnonymously = function(verb, url, data, cb) {
  describe('when called anonymously', function () {
    _beforeEach.givenAnAnonymousToken();
    _describe.whenCalledRemotely(verb, url, data, cb);
  });
}

_describe.whenCalledUnauthenticated = function(verb, url, data, cb) {
  describe('when called with unauthenticated token', function () {
    _beforeEach.givenAnAnonymousToken();
    _describe.whenCalledRemotely(verb, url, data, cb);
  });
}

_it.shouldBeAllowed = function() {
  it('should be allowed', function() {
    assert(this.req);
    assert(this.res);
    // expect success - status 2xx or 3xx
    if (this.res.statusCode < 100 || this.res.statusCode > 399) {
      console.log(this.res.body);
    }
    expect(this.res.statusCode).to.be.within(100, 399);
  });
}

_it.shouldBeDenied = function() {
  it('should not be allowed', function() {
    assert(this.res);
    var expectedStatus = this.aclErrorStatus ||
      this.app && this.app.get('aclErrorStatus') ||
      401;
      if (this.res.statusCode !== expectedStatus) {
          console.log(this.res.body);
      }
    expect(this.res.statusCode).to.equal(expectedStatus);
  });
}

_it.shouldBeRejected = function(statusCode) {
  it('should be rejected' + (statusCode ? ' with status code ' + statusCode : ''), function() {
    assert(this.res);
    if (statusCode) {
        if (this.res.statusCode !== statusCode) {
            console.log(this.res.body);
        }
        expect(this.res.statusCode).to.equal(statusCode);
    } else {
        if (this.res.statusCode < 400 || this.res.statusCode > 499) {
            console.log(this.res.body);
        }
        expect(this.res.statusCode).to.be.within(400, 499);
    }
  });
}

_it.shouldBeForbidden = function() {
  it('should be forbidden', function() {
    assert(this.res);
      if (this.res.statusCode !== 403) {
          console.log(this.res.body);
      }
    assert.equal(this.res.statusCode, 403);
  });
}

_it.shouldNotBeFound = function() {
  it('should not be found', function() {
    assert(this.res);
      if (this.res.statusCode !== 404) {
          console.log(this.res.body);
      }
    assert.equal(this.res.statusCode, 404);
  });
}

_it.shouldBeAllowedWhenCalledAnonymously =
function(verb, url, data) {
  _describe.whenCalledAnonymously(verb, url, data, function() {
    _it.shouldBeAllowed();
  });
}

_it.shouldBeDeniedWhenCalledAnonymously =
function(verb, url) {
  _describe.whenCalledAnonymously(verb, url, function() {
    _it.shouldBeDenied();
  });
}

_it.shouldBeAllowedWhenCalledUnauthenticated =
function(verb, url, data) {
  _describe.whenCalledUnauthenticated(verb, url, data, function() {
    _it.shouldBeAllowed();
  });
}

_it.shouldBeDeniedWhenCalledUnauthenticated =
function(verb, url) {
  _describe.whenCalledUnauthenticated(verb, url, function() {
    _it.shouldBeDenied();
  });
}

_it.shouldBeNotFoundWhenCalledUnauthenticated =
function(verb, url) {
  _describe.whenCalledUnauthenticated(verb, url, function() {
    _it.shouldNotBeFound();
  });
}

_it.shouldBeAllowedWhenCalledByUser =
function(credentials, verb, url, data) {
  _describe.whenCalledByUser(credentials, verb, url, data, function() {
    _it.shouldBeAllowed();
  });
}

_it.shouldBeDeniedWhenCalledByUser =
function(credentials, verb, url) {
  _describe.whenCalledByUser(credentials, verb, url, function() {
    _it.shouldBeDenied();
  });
}

_it.shouldBeAllowedWhenCalledByUserWithRole =
function(credentials, role, verb, url, data) {
  _describe.whenCalledByUserWithRole(credentials, role, verb, url, data, function() {
    _it.shouldBeAllowed();
  });
}

_it.shouldBeDeniedWhenCalledByUserWithRole =
function(credentials, role, verb, url) {
  _describe.whenCalledByUserWithRole(credentials, role, verb, url, function() {
    _it.shouldBeDenied();
  });
}

_it.shouldBeValidCreateResponse =
function () {
    _it.shouldBeAllowed();
    it('should respond with a valid POST response', function () {
        if (this.res.statusCode !== 200) {
            console.log(this.res.body);
        }
        assert.equal(this.res.statusCode, 200);
        assert(this.res.body.id);
    });
}

_it.shouldBeValidGetAllResponse =
function () {
    _it.shouldBeAllowed();
    it('should respond with a valid GET response', function () {
        if (this.res.statusCode !== 200) {
            console.log(this.res.body);
        }
        assert.equal(this.res.statusCode, 200);
        assert(Array.isArray(this.res.body));
    });
}

_it.shouldBeValidGetByIdResponse =
function (id) {
    _it.shouldBeAllowed();
    it('should respond with a valid GET response',
        function () {
            if (this.res.statusCode !== 200) {
                console.log(this.res.body);
            }
            assert.equal(this.res.statusCode, 200);
            assert.equal(this.res.body.id, id);
        });
}

_it.shouldBeValidUpdateResponse =
function (newVal) {
    _it.shouldBeAllowed();
    it('should respond with a valid PUT response',
        function () {
            if (this.res.statusCode !== 200) {
                console.log(this.res.body);
            }
            assert.equal(this.res.statusCode, 200);
            var props = Object.keys(newVal);
            var val = this.res.body;
            props.forEach(function (prop) {
                assert.equal(val[prop], newVal[prop]);
            });
        });
}

_it.shouldBeValidDeleteResponse =
function () {
    _it.shouldBeAllowed();
    it('should have statusCode 200',
        function () {
            if (this.res.statusCode !== 200) {
                console.log(this.res.body);
            }
            assert.equal(this.res.statusCode, 200);
        });
}
